Github Codespaces configuration

edwinv opened the pull request:
### Summary

Github Codespaces allows (new) Rails contributors to easily boot a fully functional environment to make and test changes to Rails. In order for this repository to work will with Codespaces, we need some configuration, this is added in this PR.

The `Dockerfile` and `boot.sh` are inspired on the [Development Dependencies Install manual](https://edgeguides.rubyonrails.org/development_dependencies_install.html) and [bootstrap.sh file](https://github.com/rails/rails-dev-box/blob/master/bootstrap.sh).

### Other Information

The devcontainer works, but has a minor issue:

In order to be able to run code right away, we need to run `bundle install` as a `postCreateCommand`. Due to the large `Gemfile`, executing this command takes long. Codespaces doesn't give proper user feedback about the time it takes, therefore a user can be confused why the environment is not fully functional. I haven't found a solution for this problem yet, suggestions welcome!

edwinv committed:
Github Codespaces configuration

edwinv committed:
Mention Codespace in contribution guide

vovimayhem commented:
Ain't better to use docker-compose with the devcontainer? downloading images for mysql, postgres, redis might be quicker and more maintainable than installing them on the Docker image...

vovimayhem commented:
Also, a entrypoint script might me added to check bundler (and yarn) dependencies... something similar to this: https://github.com/vovimayhem/rails-last-n-of-each-demo/blob/main/bin/dev-entrypoint

edwinv commented:
> Ain't better to use docker-compose with the devcontainer? downloading images for mysql, postgres, redis might be quicker and more maintainable than installing them on the Docker image...

It doesn't seem to be possible to use the `dockerComposeFile` setup in Codespaces right now. This means we need to run Docker inside the container to run Docker images. There is [a Docker library script](https://github.com/microsoft/vscode-dev-containers/blob/main/script-library/docs/docker.md) available, but is seems to be more work to add Docker support inside the container than to just install the packages from the package manager. Furthermore, the current setup is more similar to the rails-dev-box approach. When we would run dependencies in Docker, I expect changes in the configuration for connecting to the dependencies, meaning we might break things for current users.

> Also, a entrypoint script might be added to check bundler (and yarn) dependencies... something similar to this: https://github.com/vovimayhem/rails-last-n-of-each-demo/blob/main/bin/dev-entrypoint

This is handled in `boot.sh` right now.

edwinv committed:
Add bundle install and yarn install to Dockerfile

edwinv committed:
Install ImageMagick from source for latest version with SVG support

edwinv commented:
I've added the `Gemfile` and `package.json` to the Dockerfile so we can install the gems on build instead of on startup. This greatly improves the startup time for the devcontainer to be available. In the recent commit, I've added `poppler-utils` and the latest `imagemagick` version to make sure all activestorage specs pass.

The devcontainer has been tested by running the specs for all Rails components. All specs are passing, so this should be a stable workspace for development.

You can test this PR by starting a Github Codespace from the green "Edit" button in the right top corner is the screen. In the Codespace you can use the terminal to execute commands.

rafaelfranca commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,56 @@
+# [Optional] Uncomment this section to install additional OS packages.
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+    && apt-get -y install --no-install-recommends \
+        mariadb-server libmariadbclient-dev \
```
My mariadb and not mysql?

edwinv replied:
I'm more familiar with PostgreSQL, so I needed to look this up when I was writing the Dockerfile. Since [Debian 9](https://mariadb.com/kb/en/moving-from-mysql-to-mariadb-in-debian-9/), MySQL is no longer available in the default apt repositories. MariaDB is the new default database engine. [Because MariaDB is backwards compatible with MySQL and in most situations a better choice for new projects](https://stackoverflow.com/a/16187934), I assumed it isn't a problem to follow the OS default database engine approach instead of doing more work to install MySQL.

Because Github actions still run MySQL, compatibility with a real MySQL installation would still be validated when making code changes.

ghiculescu commented:
@edwinv not an expert on this, but I get the feeling more configuration is needed for this to fully work.

https://docs.github.com/en/codespaces/developing-in-codespaces/creating-a-codespace suggests it's only available to some account levels, and https://github.com/features/codespaces says this:

![image](https://user-images.githubusercontent.com/509837/134209677-ab9c41d9-7b5d-4cf6-85c8-e751747c57dd.png)

I can open a codespace by going to https://github.com/rails/rails and pressing the "." key (I couldn't see any buttons in the UI to do it), but the codespace it spins up doesn't have terminal access so I can't run tests:

![image](https://user-images.githubusercontent.com/509837/134209790-990523de-be2c-4158-a6b7-313e4c004c6b.png)

edwinv commented:
> @edwinv not an expert on this, but I get the feeling more configuration is needed for this to fully work.
> 
> https://docs.github.com/en/codespaces/developing-in-codespaces/creating-a-codespace suggests it's only available to some account levels, and https://github.com/features/codespaces says this:
>
> ...

According to the same page, access to Codespace is per repository owner and not per user. As I'm able to start Codespaces in the rails organization, I assume the organization has everything setup correctly. Personally I'm on an older Github plan that doesn't include Codespaces and I indeed have no access to it in my own repositories, but still can start Codespaces for other open source projects.

> 
> I can open a codespace by going to https://github.com/rails/rails and pressing the "." key (I couldn't see any buttons in the UI to do it), but the codespace it spins up doesn't have terminal access so I can't run tests:
> 
> ...

The `.` shortcut is merely a preview of the repository. If you want to use a Codespace, you can use the green **"Code"** button on the repo frontpage to start a Codespace:

![CleanShot 2021-09-21 at 18 59 54@2x](https://user-images.githubusercontent.com/563294/134215061-76c307c4-fc64-4d9f-b9c6-d5cc3feb0fe8.png)

ghiculescu commented:
Gotcha. Yeah, I don't have that option:

![image](https://user-images.githubusercontent.com/509837/134215661-cd3f6063-14ec-4fd1-98cf-b752544322d6.png)

edwinv commented:
> Gotcha. Yeah, I don't have that option:
> 
> ....

Weird. I interpreted from the communication from Github that it's commonly available in the mentioned plans. This seems like a Github issue and not something the repo configuration can fix.

ghiculescu commented:
Codespaces needed to be enabled for the rails org. Rafael has done that. There’s a new issue though, where we are both seeing this error when building a codespace

![image](https://user-images.githubusercontent.com/509837/134224443-6b94f7da-331c-4ce2-9c9f-56fdec44bfeb.png)

edwinv commented:
I seems like @rafaelfranca has fixed this with 7df466c3848d65d3f517c65ecce1355ad1f62025, the latest `main` works for me.

jrochkind commented:
Does codespace option under "Code" button only show up for Rails committers? I don't get the option either.
